File: MIME_outbound/MIME_retrieve.js
// Parse and convert date strings to Date objects
function parseSentDate(sentDate) {
    return new Date(sentDate);
}

// Decode base64 data
function decodeBase64(data) {
    return atob(data);
}

// Get MIME messages sent within the past 24 hours from the vehicles
function getRecentMimeMessages() {
    api.call("Get", {
        "typeName": "TextMessage",
        "resultsLimit": 10000
    }, function(result) {
        const currentTime = new Date();
        const thresholdTime = new Date(currentTime.getTime() - 24 * 60 * 60 * 1000);

        // Filter for 24 hours, MIME type, sent from vehicle
        const orderedTextMessages = result;
        const recentMessages = orderedTextMessages.filter(message => {
            const sentDate = parseSentDate(message.sent);
            return sentDate > thresholdTime &&
                   message.messageContent.contentType === 'MimeContent' &&
                   !message.isDirectionToVehicle;
        });

        // Sort and decode
        recentMessages.sort((a, b) => parseSentDate(a.sent) - parseSentDate(b.sent));
        recentMessages.forEach(message => {
            message.messageContent.decodedData = decodeBase64(message.messageContent.data);
        });
        
        // Print the sorted array of recent messages with decoded data
        console.log(recentMessages);
    }, function(e) {
        console.error("Failed:", e);
    });
}

// Call the function
getRecentMimeMessages();

File: MIME_outbound/MIME_retrieve.py
import os
import json
import pytz
import base64
import mygeotab
from getpass import getpass
from dotenv import load_dotenv
from datetime import datetime, timedelta

# Authenticate with MyGeotab to ensure user has access, then return mygeotab.API object
def authenticate_myg(username, database, server='my.geotab.com'):
    api = mygeotab.API(username=username, password=getpass('MyGeotab Password: '), database=database, server=server)
    api.authenticate()
    return api

# Fetch the TextMessage data
def fetch_text_messages(api, hours_threshold, results_limit=10000):
    text_messages = api.get(
        "TextMessage",
        resultsLimit=results_limit
    )
    text_messages_sorted = sorted(text_messages, key=lambda x: x['sent'])

    # impose filtering
    current_time = datetime.now(pytz.UTC)
    threshold_time = current_time - timedelta(hours=hours_threshold)
    response = []
    for message in text_messages_sorted[::-1]:
        if message['sent'] > threshold_time:
            if message['messageContent']['contentType'] == 'MimeContent' and not message['isDirectionToVehicle']:
                message['messageContent']['decodedData'] = base64.b64decode(message['messageContent']['data']).decode('utf-8')
                response.append(message)
        else:
            break

    return response

# Export response to JSON
def export_response(response, hours_threshold, database):
    def convert_datetime(obj):
        if isinstance(obj, datetime):
            return obj.isoformat()
        raise TypeError("Type not serializable")
    
    response = {'callParameters': {
        'database': database,
        'hoursThreshold': hours_threshold,
        'rangeStart': datetime.now(pytz.UTC),
        'rangeEnd': datetime.now(pytz.UTC) - timedelta(hours=hours_threshold),
    }, 'response': response}
    
    json_response = json.dumps(response, indent=4, default=convert_datetime)
    with open('MIME_outbound/response.json', 'w') as f:
        f.write(json_response)
    print(f"Found {len(response['response'])} messages, exported to 'MIME_outbound/response.json'.")
    
    return

def main():
    # Load environment variables from .env file
    load_dotenv()
    username = os.getenv('MYGEOTAB_USERNAME')
    database = os.getenv('MYGEOTAB_DATABASE')
    server = os.getenv('MYGEOTAB_SERVER', 'my.geotab.com')

    # Define API call params
    hours_threshold = 32*24

    api = authenticate_myg(username, database, server)
    response = fetch_text_messages(api, hours_threshold)
    export_response(response, hours_threshold, database)

    return

if __name__ == '__main__':
    main()